pkg/vm/memory/relocatable.go

```diff
@@ -6,89 +6,162 @@ import (
+//
+//   - either a Felt value (an `f.Element`),
+//   - or a pointer to another Memory Cell (a `MemoryAddress`)
+type MemoryValue struct {
+	// When isAddress is true, this indicates the segment of
+	// the memory address
+	segmentIndex uint64
+	// This represents the offset if it is an address or the
+	// value if it is a field element
+	value     *f.Element
+	isAddress bool
+}
 
-	return r, nil
+func EmptyMemoryValue() *MemoryValue {
+	return &MemoryValue{
+		value: new(f.Element),
+	}
 }
 
-// Subs two relocatables if they're in the same segment or the rhs is a Felt.
-func (r *MemoryAddress) Sub(r1 *MemoryAddress, r2 *MemoryAddress) (*MemoryAddress, error) {
-	if r2.SegmentIndex == feltSegment {
-		r.SegmentIndex = r1.SegmentIndex
-		r.Offset.Sub(&r1.Offset, &r2.Offset)
-		return r, nil
+func (mv *MemoryValue) ToMemoryAddress() (*MemoryAddress, error) {
+	if !mv.isAddress {
+		return nil, fmt.Errorf("error trying to read a memory address as field element")
 	}
+	return &MemoryAddress{
+		SegmentIndex: mv.segmentIndex,
+		Offset:       mv.value,
+	}, nil
+}
 
-	if r1.SegmentIndex != r2.SegmentIndex {
-		return nil, fmt.Errorf("cannot subtract relocatables from different segments: %d != %d", r1.SegmentIndex, r2.SegmentIndex)
+func (mv *MemoryValue) ToFieldElement() (*f.Element, error) {
+	if mv.isAddress {
+		return nil, fmt.Errorf("error trying to read a field element as a memory address")
 	}
+	return mv.value, nil
+}
 
-	r.SegmentIndex = feltSegment
-	r.Offset.Sub(&r1.Offset, &r2.Offset)
+func MemoryValueFromMemoryAddress(address *MemoryAddress) *MemoryValue {
+	return &MemoryValue{
+		segmentIndex: address.SegmentIndex,
+		value:        address.Offset,
+		isAddress:    true,
+	}
+}
 
-	return r, nil
+func MemoryValueFromFieldElement(felt *f.Element) *MemoryValue {
+	return &MemoryValue{
+		value:     felt,
+		isAddress: false,
+	}
 }
 
-// String representation of a relocatable.
-func (r MemoryAddress) String() string {
-	return fmt.Sprintf("%d:%d", r.SegmentIndex, r.Offset)
+func MemoryValueFromUint64(v uint64) *MemoryValue {
+	newElement := f.NewElement(v)
+	return &MemoryValue{
+		value: &newElement,
+	}
 }
 
-// Given a map of segment relocation, update a memory address location
-func (r *MemoryAddress) Relocate(r1 *MemoryAddress, segmentsOffsets *map[uint64]*MemoryAddress) (*MemoryAddress, error) {
-	if (*segmentsOffsets)[r1.SegmentIndex] == nil {
-		return nil, fmt.Errorf("missing segment %d relocation rule", r.SegmentIndex)
+// Adds two memory value is the second one is a Felt
+func (memVal *MemoryValue) Add(lhs *MemoryValue, rhs *MemoryValue) (*MemoryValue, error) {
+	if rhs.isAddress {
+		return nil, fmt.Errorf("cannot add two memory addresses")
 	}
 
-	r, err := r.Add((*segmentsOffsets)[r1.SegmentIndex], &MemoryAddress{0, r1.Offset})
+	fmt.Println(memVal)
+	fmt.Println("x0")
+	fmt.Println(lhs)
+	fmt.Println("x1")
+	fmt.Println(rhs)
+
+	// todo(rodro): is lhs always a memory address?
```

Comment:
I think it should be possible to add felt + memory address too

Reply:
Yes we can add it no problem, but before implementing, I'll wonder it will be useful. By implementing the first opcode I am going to have a clearer picture

Reply:
Since a cell in the memory representaion contains a MemmoryValue we won't know beforehand what value it contains. But for execution of assert opcode we'll have to perfrom arithmetic ops on them. I suggest we hide that logic into MemoryValue